components/script_plugins/unrooted_must_root.rs

```diff
@@ -139,97 +143,105 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnrootedPass {
-            // This catches assignments... the main point of this would be to catch mutable
-            // references to `JS<T>`.
-            // FIXME: Enable this? Triggers on certain kinds of uses of DomRefCell.
-            // hir::ExprAssign(_, ref rhs) => require_rooted(cx, self.in_new_function, &*rhs),
-            // This catches calls; basically, this enforces the constraint that only constructors
-            // can call other constructors.
-            // FIXME: Enable this? Currently triggers with constructs involving DomRefCell, and
-            // constructs like Vec<JS<T>> and RootedVec<JS<T>>.
-            // hir::ExprCall(..) if !self.in_new_function => {
-            //     require_rooted(cx, self.in_new_function, expr);
-            // }
-            _ => {
-                // TODO(pcwalton): Check generics with a whitelist of allowed generics.
+/// Checks if a type is unrooted or contains any owned unrooted types
+impl<'a, 'b, 'tcx> UnrootedCx<'a, 'b, 'tcx> {
+    fn is_unrooted_ty(&self, ty: &ty::TyS, in_new_function: bool) -> bool {
+        let mut ret = false;
+        let cx = self.late_cx;
+        ty.maybe_walk(|t| {
+            match t.sty {
+                ty::TyAdt(did, _) => {
+                    if cx.tcx.has_attr(did.did, "must_root") {
+                        ret = true;
+                        false
+                    } else if cx.tcx.has_attr(did.did, "allow_unrooted_interior") {
+                        false
+                    } else if match_def_path(cx, did.did, &["core", "cell", "Ref"])
+                            || match_def_path(cx, did.did, &["core", "cell", "RefMut"])
+                            || match_def_path(cx, did.did, &["core", "slice", "Iter"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "Entry"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "OccupiedEntry"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "VacantEntry"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "Iter"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "set", "Iter"]) {
+                        // Structures which are semantically similar to an &ptr.
+                        false
+                    } else if did.is_box() && in_new_function {
+                        // box in new() is okay
+                        false
+                    } else {
+                        true
+                    }
+                },
+                ty::TyParam(param_ty) => {
+                    let ty_param_def = cx.tcx.generics_of(self.def_id).type_param(&param_ty, cx.tcx);
+                    // will work for `foo::<Foo>(..)`
+                    // but won't get enough information in `bar<T>(){foo::<T>(..);}`?
```

Comment:
.

Reply:
@nikomatsakis I guess we don't care about it right know?

Reply:
I don't understand the question :)

Reply:
In normal functions we will know all of the types in MIR, but in case we are in generic functions, we can use our generic parameter (which MIR, that we have, doesn't know exact type) to call other generic functions (`bar` in example above). Then we need to recursively track the functions, I guess. Maybe I have complicated it too much. Once I'll reach "our subgoal", I'll just write some test for it and we'll see then :)

Reply:
We don't need to recursively track functions -- if we call another generic function, that function will have its `#[must_root]` annotations (or not). Example:

```rust
fn foo<#[must_root] T: Default>() { bar::<T>(); }
fn bar<U>() { }
```

Here, when we are all done, the call to `bar` will yield an error. This is because `U` is not declared with `#[must_root]`, and therefore we can't supply a value for `U` (in this case, `T`) that is "must-root".

Reply:
Oh, I see. I should have tried to write it down on a paper instead of just communicating my thought.